drivers/gyro/adxrs290/adxrs290.c

```diff
@@ -0,0 +1,324 @@
+ * @param dev - Device handler.
+ * @param hpf - Pointer to high-pass pole location container.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adxrs290_get_hpf(struct adxrs290_dev *dev, enum adxrs290_hpf *hpf)
+{
+	int32_t ret = SUCCESS;
+	uint8_t data;
+
+	ret = adxrs290_reg_read(dev, ADXRS290_REG_FILTER, &data);
+	if (IS_ERR_VALUE(ret))
+		return ret;
+
+	*hpf = ADXRS290_HPF(data);
+
+	return ret;
+}
+
+/**
+ * @brief Set the low-pass filter pole location.
+ * @param dev - Device handler.
+ * @param hpf - High-pass pole location.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adxrs290_set_hpf(struct adxrs290_dev *dev, enum adxrs290_hpf hpf)
+{
+	int32_t ret = SUCCESS;
+	uint8_t reg_val;
+
+	ret = adxrs290_reg_read(dev, ADXRS290_REG_FILTER, &reg_val);
+	if (IS_ERR_VALUE(ret))
+		return ret;
+
+	reg_val &= ~ADXRS290_HPF_MASK;
+	reg_val |= (hpf << 4) & ADXRS290_HPF_MASK;
+
+	return adxrs290_reg_write(dev, ADXRS290_REG_FILTER, reg_val);
+}
+
+/**
+ * @brief Get the Gyro data channels.
+ * @param dev - Device handler.
+ * @param ch - Channel to read.
+ * @param rate - Pointer to rate value.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adxrs290_get_rate_data(struct adxrs290_dev *dev,
+			       enum adxrs290_channel ch, int16_t *rate)
+{
+	int32_t ret = SUCCESS;
+	uint8_t data[3];
+
+	data[0] = 0x80 | (ADXRS290_REG_DATAX0+ch*2);
+	data[1] = 0x80 | (ADXRS290_REG_DATAX1+ch*2);
+	data[2] = 0;
+	ret = spi_write_and_read(dev->spi_desc, data, 3);
+	if (IS_ERR_VALUE(ret))
+		return ret;
+
+	*rate = (((int16_t)data[2]) << 8) | data[1];
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Get the temperature data.
+ * @param dev - Device handler.
+ * @param temp - Pointer to temperature value.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adxrs290_get_temp_data(struct adxrs290_dev *dev, int16_t *temp)
+{
+	int32_t ret = SUCCESS;
+	uint8_t data[3];
+
+	data[0] = 0x80 | ADXRS290_REG_TEMP0;
+	data[1] = 0x80 | ADXRS290_REG_TEMP1;
+	data[2] = 0;
+
+	ret = spi_write_and_read(dev->spi_desc, data, 3);
+	if (IS_ERR_VALUE(ret))
+		return ret;
+
+	*temp = ((((int16_t)data[2]) << 8) | data[1]) & 0x0FFF;
+	*temp = (*temp << 4) >> 4;
+	return ret;
+}
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ *		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adxrs290_init(struct adxrs290_dev **device,
+		      const struct adxrs290_init_param *init_param)
+{
+	struct adxrs290_dev *dev;
+	int32_t ret = 0;
+	uint8_t val = 0;
+
+	dev = (struct adxrs290_dev *)malloc(sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	ret = spi_init(&dev->spi_desc, &init_param->spi_init);
+	if (IS_ERR_VALUE(ret))
+		goto error;
+
+	ret = adxrs290_reg_read(dev, ADXRS290_REG_DEV_ID, &val);
+	if (IS_ERR_VALUE(ret) || (val != ADXRS290_DEV_ID))
+		goto error;
```

Comment:
I would suggest making two error labels: `error_dev` and `error_spi`.
After `error_dev` call `free(dev);` as you do now and after `error_spi` call `spi_remove(dev->spi_desc);`.
Make the `spi_init(&dev->spi_desc, &init_param->spi_init);` error go to `error_dev` and the rest to `error_spi`.
That way we avoid memory leak from `dev->spi_desc` being allocated and never freed  before `dev` is gone.

Reply:
Thanks. Yeah, I missed this one. I implemented your suggestion.